### PR TITLE
[MESOS][SPARK-8873] Clean up external shuffle data on driver exit with Mesos.

### PR DESCRIPTION
**Pull request by tnachen**
@andrewor14 @dragos

### CODE REVIEW

**Comment by andrewor14** [comment]
retest this please

**Comment by andrewor14** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        blockHandler.applicationRemoved(connectedApps(remoteAddress), true)
+        connectedApps.remove(remoteAddress)
+      } else {
+        logWarning(s"Address $remoteAddress not found in mesos shuffle service")
+      }
+    }
+  }
+}
+
+case class RegisterMesosDriver(appId: String, address: RpcAddress)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala]: private

File: core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+import org.apache.spark.{Logging, SecurityManager, SparkConf}
+
+private[mesos] class MesosExternalShuffleService(conf: SparkConf, securityManager: SecurityManager)
+  extends ExternalShuffleService(conf, securityManager) {
+
+  private var rpcEnv: RpcEnv = _
+
+  override def start(): Unit = {
+    super.start()
+    val port = conf.getInt("spark.mesos.shuffle.service.port", 7327)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala]: no need to make a new config for this one, just reuse `spark.shuffle.service.port` at 7337

File: core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+import java.util.concurrent.CountDownLatch
+
+import scala.collection.mutable
+
+import org.apache.spark.deploy.ExternalShuffleService
+import org.apache.spark.rpc.{RpcAddress, RpcEnv, ThreadSafeRpcEndpoint}
+import org.apache.spark.util.Utils
+import org.apache.spark.{Logging, SecurityManager, SparkConf}
+
+private[mesos] class MesosExternalShuffleService(conf: SparkConf, securityManager: SecurityManager)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala]: This is missing a top level javadoc. You need to explain why Mesos needs a special shuffle service, i.e. it needs to detect when drivers exit, and the way we do it is through a separate RPC channel between the driver and each shuffle service.

File: core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+  }
+
+  private class MesosExternalShuffleServiceEndpoint(override val rpcEnv: RpcEnv)
+    extends ThreadSafeRpcEndpoint with Logging {
+
+    // Stores a map of driver rpc addresses to app ids
+    private val connectedApps = new mutable.HashMap[RpcAddress, String]
+
+    override def receive: PartialFunction[Any, Unit] = {
+      case RegisterMesosDriver(appId, address) =>
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala]: This isn't a Mesos driver right? This is a Spark driver. I would just call this `RegisterDriver`

File: core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    rpcEnv.setupEndpoint(MesosExternalShuffleService.ENDPOINT_NAME, endpoint)
+  }
+
+  override def stop(): Unit = {
+    rpcEnv.shutdown()
+    rpcEnv.awaitTermination()
+    super.stop()
+  }
+
+  private class MesosExternalShuffleServiceEndpoint(override val rpcEnv: RpcEnv)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala]: can you add a java doc for this class:
```
An RPC endpoint for talking to the Spark driver.
```

File: core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        if (connectedApps.contains(address)) {
+          val existingAppId: String = connectedApps(address)
+          if (!existingAppId.equals(appId)) {
+            logError(s"A new app id $appId has connected to existing address $address" +
+                     s", removing registered app $existingAppId")
+            blockHandler.applicationRemoved(existingAppId, true)
+          }
+        } else {
+          connectedApps(address) = appId
+        }
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala]: should this be an else case? In L68 we removed the existing application ID if there's a duplicate. Shouldn't we put the updated one in `connectedApps`?

File: core/src/main/scala/org/apache/spark/rpc/RpcEndpoint.scala [hunk 7]
```diff
@@ -44,14 +44,14 @@ private[spark] trait ThreadSafeRpcEndpoint extends RpcEndpoint
 
 /**
  * An end point for the RPC that defines what functions to trigger given a message.
- *
+ *M
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/rpc/RpcEndpoint.scala]: extra M

File: core/src/main/scala/org/apache/spark/rpc/RpcEndpoint.scala [hunk 8]
```diff
@@ -44,14 +44,14 @@ private[spark] trait ThreadSafeRpcEndpoint extends RpcEndpoint
...
  * It is guaranteed that `onStart`, `receive` and `onStop` will be called in sequence.
  *
- * The lift-cycle will be:
+ * The life-cycle of an end point is:
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/rpc/RpcEndpoint.scala]: endpoint 1 word

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala [hunk 9]
```diff
@@ -244,6 +248,10 @@ private[spark] class CoarseMesosSchedulerBackend(
 
           // accept the offer and launch the task
           logDebug(s"Accepting offer: $id with attributes: $offerAttributes mem: $mem cpu: $cpus")
+          if (conf.getBoolean("spark.shuffle.service.enabled", false)) {
+            slaveIdsToShuffleEndpoints(offer.getSlaveId.getValue) =
+              RpcAddress(offer.getHostname, conf.getInt("spark.mesos.shuffle.service.port", 7327))
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala]: same here, just use old port and config

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala [hunk 10]
```diff
@@ -62,6 +64,8 @@ private[spark] class CoarseMesosSchedulerBackend(
 
   val slaveIdsWithExecutors = new HashSet[String]
 
+  val slaveIdsToShuffleEndpoints = new HashMap[String, RpcAddress]
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala]: this needs a comment: `// Only used if shuffle service is enabled`

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala [hunk 11]
```diff
@@ -261,7 +269,27 @@ private[spark] class CoarseMesosSchedulerBackend(
...
+          MesosExternalShuffleService.SYSTEM_NAME,
+          rpcAddress,
+          MesosExternalShuffleService.ENDPOINT_NAME).onComplete { ref =>
+          if (ref.isSuccess) {
+            ref.get.send(RegisterMesosDriver(conf.getAppId, sc.env.rpcEnv.address))
+          } else {
+            logWarning(s"Unable to connect to shuffle service for slave ${slaveId}" +
+              s", please manually remove shuffle data after driver exit. " +
+              s"Error: ${ref.failed.get}")
+          }
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala]: these blocks need to be indented

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala [hunk 12]
```diff
@@ -261,7 +269,27 @@ private[spark] class CoarseMesosSchedulerBackend(
...
+        slaveIdsToShuffleEndpoints.contains(slaveId)) {
+        val rpcAddress = slaveIdsToShuffleEndpoints.remove(slaveId).get
+        logDebug(s"Connecting to shuffle service on slave ${slaveId}, " +
+            s"address $rpcAddress for app ${conf.getAppId}")
+        sc.env.rpcEnv.asyncSetupEndpointRef(
+          MesosExternalShuffleService.SYSTEM_NAME,
+          rpcAddress,
+          MesosExternalShuffleService.ENDPOINT_NAME).onComplete { ref =>
+          if (ref.isSuccess) {
+            ref.get.send(RegisterMesosDriver(conf.getAppId, sc.env.rpcEnv.address))
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala]: is it safe to just call `get` here? Should we do a null check just in case? Same in L288

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala]: also nit: you use `sc.env.rpcEnv` in a few places, can you make it a local variable `val rpcEnv`

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala [hunk 13]
```diff
@@ -261,7 +269,27 @@ private[spark] class CoarseMesosSchedulerBackend(
     val taskId = status.getTaskId.getValue.toInt
     val state = status.getState
     logInfo(s"Mesos task $taskId is now $state")
+    val slaveId: String = status.getSlaveId.getValue
     stateLock.synchronized {
+      if (TaskState.fromMesos(state).equals(TaskState.RUNNING) &&
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala]: this block of code needs a huge comment:
```
// If the shuffle service is enabled, have the driver register with each one
// of them. This is important for cleaning up state on the shuffle services
// after this application exits; there is currently not a great way to detect
// this through Mesos, since the shuffle services are set up independently.
```

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala [hunk 14]
```diff
@@ -261,7 +269,27 @@ private[spark] class CoarseMesosSchedulerBackend(
...
+        slaveIdsToShuffleEndpoints.contains(slaveId)) {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala]: indent 2 spaces

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala (hunk 12 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala]: also, should we do an `askWithRetry` here? Right now this is a one-way message and we have no idea whether this is actually received. More concretely I'm suggesting that you add a `DriverRegistered` response message sent from the shuffle service.

File: sbin/start-mesos-shuffle-service.sh [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# Starts the external shuffle server on the machine this script is executed on.
+#
+# Usage: start-shuffle-server.sh
```
Comment (andrewor14) [inline sbin/start-mesos-shuffle-service.sh]: outdated, the comment up there is also outdated

Comment (andrewor14) [inline sbin/start-mesos-shuffle-service.sh]: also, please add in the comment here that "this shuffle service detects when an application exits and automatically cleans up its shuffle files", and the fact that this is hard to do through Mesos.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala [hunk 16]
```diff
@@ -261,7 +269,27 @@ private[spark] class CoarseMesosSchedulerBackend(
...
+        val rpcAddress = slaveIdsToShuffleEndpoints.remove(slaveId).get
+        logDebug(s"Connecting to shuffle service on slave ${slaveId}, " +
+            s"address $rpcAddress for app ${conf.getAppId}")
+        sc.env.rpcEnv.asyncSetupEndpointRef(
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala]: why async? just wondering

File: core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala (hunk 1 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala]: also please add a simple java doc, even something like `A message sent from the driver to register with this shuffle service.` would do.

File: core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    override def onDisconnected(remoteAddress: RpcAddress): Unit = {
+      logDebug(s"Received disconnect from $remoteAddress")
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala]: this should be at least `logInfo`. Also I would rephrase this:
```
s"Application $appId disconnected (address was $remoteAddress)"
```

File: core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+    super.stop()
+  }
+
+  private class MesosExternalShuffleServiceEndpoint(override val rpcEnv: RpcEnv)
+    extends ThreadSafeRpcEndpoint with Logging {
+
+    // Stores a map of driver rpc addresses to app ids
+    private val connectedApps = new mutable.HashMap[RpcAddress, String]
+
+    override def receive: PartialFunction[Any, Unit] = {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala]: as mentioned below, this would become `receiveAndReply`

**Comment by andrewor14** [comment]
@tnachen looks great. My comments are mostly minor but I think this is very close.

File: core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala (hunk 2 above)
Comment (tnachen) [inline core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala]: It won't clash on the same port?

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala]: why would you have both shuffle services on the same cluster? Your new shuffle service is a replacement for the old one, no?

Comment (tnachen) [inline core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala]: It's actually a additional endpoint on a different port that is listening for the driver.

Let me take a look if I can easily manage to use the same, but it looks like it's not using the endpoint mechanism.

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala]: oh I see, currently 7337 is for executors while 7327 is for the driver. Yeah I think it would be very valuable to avoid another config.

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala]: One thing you could do is have a `MesosExternalShuffleBlockHandler` that extends `ExternalShuffleBlockHandler`. It would override the `receive` method to also accept a `RegisterDriver` message. Note that because it extends `RpcHandler` there is also a `connectionTerminated` method, so we can try using that one instead.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala (hunk 16 above)
Comment (tnachen) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala]: Didn't feel like this needs to be blocking since we need the results, so just ended up doing a async call instead.

File: sbin/start-mesos-shuffle-service.sh (hunk 15 above)
Comment (tnachen) [inline sbin/start-mesos-shuffle-service.sh]: I added a TODO in the CoarseMesosSchedulerBackend

File: core/src/main/scala/org/apache/spark/deploy/ExternalShuffleService.scala [hunk 19]
```diff
@@ -37,15 +39,17 @@ import org.apache.spark.util.Utils
  * Optionally requires SASL authentication in order to read. See [[SecurityManager]].
  */
 private[deploy]
-class ExternalShuffleService(sparkConf: SparkConf, securityManager: SecurityManager)
+class ExternalShuffleService(
+    sparkConf: SparkConf,
+    securityManager: SecurityManager,
+    transportConf: TransportConf,
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/ExternalShuffleService.scala]: You don't need to pass this in here. In fact, do we even use this, since we create another one from `sparkConf` down there?

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/ExternalShuffleService.scala]: This interface is confusing: we pass in both a `SparkConf` and a `TransportConf`, even though one is created from another one.

File: core/src/main/scala/org/apache/spark/deploy/ExternalShuffleService.scala [hunk 20]
```diff
@@ -37,15 +39,17 @@ import org.apache.spark.util.Utils
...
+    blockHandler: ExternalShuffleBlockHandler)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/ExternalShuffleService.scala]: I would do this a different way, just define a
```
protected def newBlockHandler(): ExternalShuffleBlockHandler = {
  new ExternalShuffleBlockHandler(transportConf)
}
```
and then have the subclass override it

File: core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala [hunk 21]
```diff
@@ -0,0 +1,142 @@
...
+import org.apache.spark.util.Utils
+import org.apache.spark.{Logging, SecurityManager, SparkConf}
+
+import scala.collection.mutable
+
+
+/**
+ * MesosExternalShuffleServiceEndpoint is a RPC endpoint that receives
+ * registration requests from Spark drivers launched with Mesos.
+ * It detects driver termination and calls the cleanup callback to [[ExternalShuffleService]]
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala]: nit: just start with `An RPC endpoint that...`

File: core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala [hunk 22]
```diff
@@ -0,0 +1,142 @@
...
+import org.apache.spark.network.shuffle.protocol.mesos.RegisterDriver
+import org.apache.spark.network.util.TransportConf
+import org.apache.spark.util.Utils
+import org.apache.spark.{Logging, SecurityManager, SparkConf}
+
+import scala.collection.mutable
+
+
+/**
+ * MesosExternalShuffleServiceEndpoint is a RPC endpoint that receives
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala]: nit: just start with `An RPC endpoint that...`

File: core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala [hunk 23]
```diff
@@ -0,0 +1,142 @@
...
+    message match {
+      case RegisterDriverParam(appId) =>
+        val address = client.getSocketAddress()
+        logDebug(s"Received registration request from app $appId, address $address")
+        if (connectedApps.contains(address)) {
+          val existingAppId: String = connectedApps(address)
+          if (!existingAppId.equals(appId)) {
+            logError(s"A new app id $appId has connected to existing address $address" +
+              s", removing registered app $existingAppId")
+            applicationRemoved(existingAppId, true)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala]: can you name this boolean parameter

File: core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala [hunk 24]
```diff
@@ -0,0 +1,142 @@
...
+    } else {
+      logWarning(s"Address $address not found in mesos shuffle service")
+    }
+  }
+}
+
+/**
+ * An extractor object for matching RegisterDriver message.
+ */
+private[mesos] object RegisterDriverParam {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala]: this could be strictly `private`

File: core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala [hunk 25]
```diff
@@ -0,0 +1,142 @@
...
+  def unapply(r: RegisterDriver): Option[String] = Some(r.getAppId())
+}
+
+/**
+ * MesosExternalShuffleService wraps [[ExternalShuffleService]] which provides an additional
+ * endpoint for drivers to associate with. This allows the shuffle service to detect when
+ * a driver is terminated and can further clean up the cached shuffle data.
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala]: technically incorrect: shuffle data is not cached. Just say the `clean up the shuffle files`

File: core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala [hunk 26]
```diff
@@ -0,0 +1,142 @@
...
+    conf: SparkConf,
+    securityManager: SecurityManager,
+    transportConf: TransportConf)
+  extends ExternalShuffleService(
+    conf, securityManager, transportConf, new MesosExternalShuffleBlockHandler(transportConf)) {
+}
+
+private[spark] object MesosExternalShuffleService extends Logging {
+  val SYSTEM_NAME = "mesosExternalShuffleService"
+  val ENDPOINT_NAME = "mesosExternalShuffleServiceEndpoint"
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala]: not used, please clean up the code after an update

File: core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala [hunk 27]
```diff
@@ -0,0 +1,142 @@
...
+ */
+private[mesos] class MesosExternalShuffleService(
+    conf: SparkConf,
+    securityManager: SecurityManager,
+    transportConf: TransportConf)
+  extends ExternalShuffleService(
+    conf, securityManager, transportConf, new MesosExternalShuffleBlockHandler(transportConf)) {
+}
+
+private[spark] object MesosExternalShuffleService extends Logging {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala]: this whole thing is duplicated from `ExternalShuffleService`. We should at least attempt to reuse some of the code... I think if you make this object inherit the `ExternalShuffleService` object we can get a lot done.

File: core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala [hunk 28]
```diff
@@ -17,28 +17,26 @@
...
+import org.apache.spark.network.shuffle.ExternalShuffleBlockHandler
 import org.apache.spark.rpc._
-import org.apache.spark.util.{ThreadUtils, SignalLogger, Utils}
+import org.apache.spark.util.{SignalLogger, ThreadUtils, Utils}
+import org.apache.spark.{Logging, SecurityManager, SparkConf}
+
+import scala.collection.mutable.{HashMap, HashSet, LinkedHashMap}
+import scala.concurrent.ExecutionContext
+import scala.util.Random
+import scala.util.control.NonFatal
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala]: these imports should be organized

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala]: actually this file doesn't need to change if we don't modify the constructor as suggested above

File: network/shuffle/src/main/java/org/apache/spark/network/shuffle/mesos/MesosExternalShuffleClient.java [hunk 29]
```diff
@@ -0,0 +1,63 @@
...
+import org.apache.spark.network.client.RpcResponseCallback;
+import org.apache.spark.network.client.TransportClient;
+import org.apache.spark.network.sasl.SecretKeyHolder;
+import org.apache.spark.network.shuffle.ExternalShuffleClient;
+import org.apache.spark.network.shuffle.protocol.mesos.RegisterDriver;
+import org.apache.spark.network.util.TransportConf;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MesosExternalShuffleClient extends ExternalShuffleClient {
```
Comment (andrewor14) [inline network/shuffle/src/main/java/org/apache/spark/network/shuffle/mesos/MesosExternalShuffleClient.java]: this needs a huge java doc

**Comment by andrewor14** [comment]
@tnachen Thanks for addressing the comments quickly. The latest changes need a bit of clean up, however. Have you had a chance to test the latest changes on a real cluster and verify that the shuffle files did get cleaned up?

**Comment by tnachen** [comment]
Yes I tested with a single master and slave and it worked.
I can get to the comments after 7 today.

File: core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala (hunk 27 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala]: never mind you can't extend objects.

**Comment by andrewor14** [comment]
Hey @tnachen I opened #7881 after addressing the comments myself. Could you test out the latest changes in that patch instead?

**Comment by andrewor14** [comment]
Can you close this one now that #7881 is already merged?

**Comment by tnachen** [comment]
This is already merged.

File: sbin/start-mesos-shuffle-service.sh [hunk 30]
```diff
@@ -0,0 +1,35 @@
...
+# Use the SPARK_SHUFFLE_OPTS environment variable to set shuffle server configuration.
+#
+
+sbin="`dirname "$0"`"
+sbin="`cd "$sbin"; pwd`"
+
+. "$sbin/spark-config.sh"
+. "$SPARK_PREFIX/bin/load-spark-env.sh"
+
+exec "$sbin"/spark-daemon.sh start org.apache.spark.deploy.mesos.MesosExternalShuffleService 1
```
Comment (dragos) [inline sbin/start-mesos-shuffle-service.sh]: Do we still need the old one, `start-shuffle-service.sh`? It was only used by Mesos as far as I remember.

Comment (andrewor14) [inline sbin/start-mesos-shuffle-service.sh]: I think we need to keep it for backward compatibility unfortunately.
